README.md
Mojo-Weixin v1.4.6 [![Build Status](https://travis-ci.org/sjdy521/Mojo-Weixin.svg?branch=master)](https://travis-ci.org/sjdy521/Mojo-Weixin) [![API列表](https://img.shields.io/badge/doc-API%E5%88%97%E8%A1%A8-brightgreen.svg)](API.md) [![QQ群-498880156](https://img.shields.io/badge/QQ%E7%BE%A4-498880156-blue.svg)](http://jq.qq.com/?_wv=1027&k=kjVJzo) [![IRC](https://img.shields.io/badge/IRC-%E5%AE%98%E6%96%B9%E9%A2%91%E9%81%93-yellow.svg)](http://irc.perfi.wang/?channel=#Mojo-Webqq)
[![LICENSE](https://img.shields.io/badge/license-Anti%20996-blue.svg)](https://github.com/996icu/996.ICU/blob/master/LICENSE)
========================

使用Perl语言编写的微信客户端框架，基于Mojolicious，要求Perl版本5.10+，可通过插件提供基于HTTP协议的api接口供其他语言或系统调用

### 郑重声明

本项目完全遵循微信官方提供的原始功能，不包含任何破坏、盗号等行为，本项目完全开源，目的是为了促进技术交流学习，禁止任何对国家、企业、个人等构成危害的非法用途，否则后果自负

### 特色功能

* 支持 发送/接收文字、图片、视频、语音、表情、文件等多种媒体消息（网页版自身功能范围）
* 支持 新增/失去好友、新增/退出群组、新增/失去群成员 等事件提醒
* 支持 创建群组、群组加人/踢人、发送/接受好友验证申请、设置群组名称、设置好友备注
* 支持 Windows/Linux/Mac 多平台，支持docker镜像，易安装部署，不懂Perl也能用
* 提供 基于HTTP协议的API接口 ，简洁丰富，方便和其他编程语言集成
* 一个产品经理的倾情之作，代码由内而外都更加注重“用户体验”，外加主人人品极好，你会懂的

### 插件列表

|名称                | 优先级  |当前状态    |github作者   | 功能说明                 
|:-------------------|:--------|:-----------|:------------|:----------------------------
|[ShowMsg](https://metacpan.org/pod/distribution/Mojo-Weixin/lib/Mojo/Weixin.pod#Mojo::Weixin::Plugin::ShowMsg)             |100      |已发布      |sjdy521      |打印客户端接收和发送的消息
|[IRCShell](https://metacpan.org/pod/distribution/Mojo-Weixin/lib/Mojo/Weixin.pod#Mojo::Weixin::Plugin::IRCShell)            |99       |已发布      |sjdy521      |Linux环境下通过irc客户端使用微信
|[Openwx](https://metacpan.org/pod/distribution/Mojo-Weixin/lib/Mojo/Weixin.pod#Mojo::Weixin::Plugin::Openwx)              |98       |已发布      |sjdy521      |提供微信发送消息api接口
|[Beauty](https://metacpan.org/pod/distribution/Mojo-Weixin/lib/Mojo/Weixin.pod#Mojo::Weixin::Plugin::Beauty)              |95       |已发布      |sjdy521      |识别指定关键字发送美女图片
|[Translation](https://metacpan.org/pod/distribution/Mojo-Weixin/lib/Mojo/Weixin.pod#Mojo::Weixin::Plugin::Translation)         |93       |已发布      |sjdy521      |多国语言翻译功能
|[KnowledgeBase](https://metacpan.org/pod/distribution/Mojo-Weixin/lib/Mojo/Weixin.pod#Mojo::Weixin::Plugin::KnowledgeBase)       |2        |已发布      |sjdy521      | 通过微信消息自定义问答知识库
|[AutoVerify](https://metacpan.org/pod/distribution/Mojo-Weixin/lib/Mojo/Weixin.pod#Mojo::Weixin::Plugin::AutoVerify)          |1        |已发布      |sjdy521      |收到好友验证请求时自动批准同意
|[PostQRcode](https://metacpan.org/pod/distribution/Mojo-Weixin/lib/Mojo/Weixin.pod#Mojo::Weixin::Plugin::PostQRcode)          |0        |已发布      |sjdy521      |登录二维码发送到邮箱实现远程扫码
|[UploadQRcode](https://metacpan.org/pod/distribution/Mojo-Weixin/lib/Mojo/Weixin.pod#Mojo::Weixin::Plugin::UploadQRcode)          |0        |已发布      |sjdy521      |二维码上传图床获得公网访问url
|[XiaoiceReply](https://metacpan.org/pod/distribution/Mojo-Weixin/lib/Mojo/Weixin.pod#Mojo::Weixin::Plugin::XiaoiceReply)        |1        |已发布      |sjdy521      |利用微软小冰实现智能聊天回复
|[ShowQRcode](https://metacpan.org/pod/distribution/Mojo-Weixin/doc/Weixin.pod#Mojo::Weixin::Plugin::ShowQRcode)          |0        |已发布      |sjdy521  |调用系统图片查看程序来示二维码（目前仅支持win）
|[SmartReply](https://metacpan.org/pod/distribution/Mojo-Weixin/lib/Mojo/Weixin.pod#Mojo::Weixin::Plugin::SmartReply)          |0        |已发布      |sjdy521      |智能聊天回复

更多插件参考[完整插件列表](Plugin.md)

### 效果展示

敲一行命令就能启动一个智能聊天机器人，Perl 和你都如此优雅. Enjoy!

    cpanm Mojo::Weixin && perl -MMojo::Weixin -e "Mojo::Weixin->new->load('ShowMsg')->load('SmartReply')->run()"
    
```
[16/01/19 23:10:49] [info] 客户端准备登录...
[16/01/19 23:10:50] [info] 清除残留的历史二维码图片
[16/01/19 23:10:30] [info] 二维码已下载到本地[ /tmp/mojo_weixin_qrcode.jpg ]
[16/01/19 23:11:20] [info] 等待手机微信扫描二维码...
[16/01/19 23:12:09] [info] 手机微信扫码成功，请在手机微信上点击 [登录] 按钮...
[16/01/19 23:12:10] [info] 正在进行登录...
[16/01/19 23:12:10] [info] 微信登录成功
[16/01/19 23:13:40] [info] 获取联系人信息...
[16/01/19 23:13:40] [info] 更新个人信息成功
[16/01/19 23:14:41] [info] 更新好友信息成功
[16/01/19 23:15:42] [info] 更新群组[ 红包群 ]信息成功
[16/01/19 23:15:42] [info] 更新群组[ Perl语言交流 ]信息成功
[16/01/19 23:15:42] [info] 开始接收消息...
[16/01/19 23:15:00] [群消息] 小灰|Perl语言交流 : Mojo::Weixin不错哦
[16/01/19 23:15:58] [群消息] 我->Perl语言交流 : 多谢多谢
```
#### 通过irc客户端在linux终端上使用微信

```
    +-------------------+                      +----------------+  
    |  Tencent          |                      | Any IRC Client |
    |  Weixin Server    |                      | wechat、irssi  |
    +---v-------------^-+                      +-v------------^-+     
        |             |                          |            |
        | 微信协议交互|                          |IRC协议交互 |
+-- --- |--  - -  --  | - - -   --   -  -   ---  | ---  ----- | --+
|   +---v-------------^--+                  +----v------------^-+ |   
|   |                    <——————————————————<                   | |
|   |   Weixin  Client   |  Weixin - IRC    |  IRC Server       | |
|   |                    |     协议转换     |  监听本机6667端口 | |
|   |                    >——————————————————>                   | | 
|   +--------------------+                  +-------------------+ |
|                                                                 |
|                                       我们程序实现的部分        | 
+---  - - - -  -- - --  ----  ------  -------  ------  ---    ----+

```

![IRCShell](screenshot/IRCShell.jpg)

#### 更多应用效果 请参见[Mojo-Weixin作品展](Collection.md)

### 安装方法

推荐使用[cpanm](https://metacpan.org/pod/distribution/App-cpanminus/bin/cpanm)在线安装[Mojo::Weixin](https://metacpan.org/pod/distribution/Mojo-Weixin/doc/Weixin.pod)模块, 如果使用docker方式请参见[Docker镜像安装及使用方法](Docker.md)

1. *安装perl*
  
    安装之前请先确认下你的系统是否已经安装了Perl，因为除了windows，其他大部分的平台默认都可能已经预装过

    并且你的Perl版本至少5.10.1+，推荐5.14+

    [Perl官网下载页面](https://www.perl.org/get.html) 有包含Unix/Linux、Mac OS X、Windows多平台比较全面详细的安装说明

    建议大家尽量选择**Binaries**（二进制预编译）的版本，安装即可使用，比较方便

    Linux上如果你的perl版本较低，想要升级perl，可以考虑使用工具[perlbrew](https://perlbrew.pl/)

    |平台    |推荐选择  |下载地址         |
    |:-------|:--------|:---------------|
    |Windows |1. **StrawberryPerl**<br>2. ActivePerl<br>3. **Mojo-StrawberryPerl**|[StrawberryPerl下载地址](http://strawberryperl.com/)<br>[Mojo-StrawberryPerl下载地址](https://github.com/sjdy521/Mojo-StrawberryPerl)<br>[ActivePerl下载地址](http://www.activestate.com/activeperl/downloads)<br>|
    |Linux   |1. **系统自带**<br>2. **yum/apt等包管理器**<br>3. 官方源码编译<br>4. ActivePerl<br>5. DWIM-Perl|[ActivePerl下载地址](http://www.activestate.com/activeperl/downloads)<br>[DWIM-Perl下载地址](http://dwimperl.com/linux.html)|
    |Mac     |1. **系统自带**<br>2. ActivePerl|[ActivePerl下载地址](http://www.activestate.com/activeperl/downloads)|
  
    注意：
    
    [Mojo-StrawberryPerl](https://github.com/sjdy521/Mojo-StrawberryPerl)是基于StrawberryPerl精简打包而成，适用于windows 32位/64位系统

2. *安装cpanm工具*（如果系统已经安装了cpanm可以忽略此步骤）

    方法a： 通过cpan安装cpanm（只适用于root用户）

        $ cpan -i App::cpanminus
    
    方法b： 直接在线安装cpanm

        $ curl -kL http://cpanmin.us | perl - App::cpanminus

        官方服务器在国外，如果无法访问，可以选择下面的命令：
        
        $ curl http://share-10066126.cos.myqcloud.com/cpanm.pl|perl - App::cpanminus

2. *使用cpanm在线安装 Mojo::Weixin 模块*（如果系统已经安装了该模块，执行此步骤会对模块进行升级）

        $ cpanm Mojo::Weixin
    
    如果安装过程中一直提示下载失败，很可能是因为访问到国外服务器网络比较差
    
    这种情况下可以尝试按如下方式手动指定国内的镜像站点
    
        $ cpanm --mirror http://mirrors.163.com/cpan/ Mojo::Weixin

3. *安装失败可能有帮助的解决方法*
        
    如果你运气不佳，通过cpanm没有一次性安装成功，这里提供了一些可能有用的信息

    在安装 Mojo::Weixin 的过程中，cpan或者cpanm会帮助我们自动安装很多其他的依赖模块
    
    在众多的依赖模块中，安装经常容易出现问题的主要是 IO::Socket::SSL
    
    IO::Socket::SSL 主要提供了 https 支持，在安装过程中可能会涉及到SSL相关库的编译

    对于 Linux 用户，通常采用的是编译安装的方式，系统缺少编译安装必要的环境，则会导致编译失败
    
    对于 Windows 用户，由于不具备良好的编译安装环境，推荐采用一些已经打包比较全面的Perl运行环境
    
    例如比较流行的 strawberryperl 或者 activeperl 的最新版本都默认包含 Mojo::Weixin 的核心依赖模块

    RedHat/Centos:

        $ yum install -y openssl-devel
        
    Ubuntu:

        $ sudo apt-get install libssl-dev
    
    搞定了编译和运行环境之后，再重新回到 步骤2 安装Mojo::Weixin即可
    
    请尽量在root下安装使用，非root下安装使用参考[FAQ](https://github.com/sjdy521/Mojo-Weixin/blob/master/FAQ.md#5-%E9%9D%9Eroot%E8%B4%A6%E5%8F%B7%E5%AE%89%E8%A3%85%E5%90%8E%E6%97%A0%E6%B3%95%E4%BD%BF%E7%94%A8%E9%97%AE%E9%A2%98)
    
    如果仍然有问题，Linux上可以尝试执行如下脚本进行依赖模块检查
    
        $ curl -ks "https://raw.githubusercontent.com/sjdy521/Mojo-Weixin/master/script/check_dependencies.pl"|perl -
        

### 如何使用

1. *我对Perl很熟悉，是一个专业的Perler*

    该项目是一个纯粹的Perl模块，已经发布到了cpan上，请仔细阅读 `Mojo::Weixin` 模块的[使用文档](https://metacpan.org/pod/distribution/Mojo-Weixin/doc/Weixin.pod)

    除此之外，你可以看下 [demo](https://github.com/sjdy521/Mojo-Weixin/tree/master/demo) 目录下的更多代码示例

2. *我是对Perl不熟悉，是一个其他语言的开发者，只对提供的消息发送/接收接口感兴趣*

    可以直接把如下代码保存成一个源码文件(必须使用UTF8编码)，使用 perl 解释器来运行
    
        #!/usr/bin/env perl
        use Mojo::Weixin;
        my ($host,$port,$post_api);
        
        $host = "0.0.0.0"; #发送消息接口监听地址，没有特殊需要请不要修改
        $port = 3000;      #发送消息接口监听端口，修改为自己希望监听的端口
        #$post_api = 'http://xxxx';  #接收到的消息上报接口，如果不需要接收消息上报，可以删除或注释此行
        
        my $client = Mojo::Weixin->new(log_level=>"info",http_debug=>0);
        $client->load("ShowMsg");
        $client->load("Openwx",data=>{listen=>[{host=>$host,port=>$port}], post_api=>$post_api});
        $client->run();
    
    上述代码保存成 xxxx.pl 文件，然后使用 perl 来运行，就会完成 微信 登录并在本机产生一个监听指定地址端口的 http server
    
        $ perl xxxx.pl
    
    发送好友消息的接口调用示例
    
        http://127.0.0.1:3000/openwx/send_friend_message?id=xxxx&content=hello
        
        * About to connect() to 127.0.0.1 port 3000 (#0)
        *   Trying 127.0.0.1...
        * Connected to 127.0.0.1 (127.0.0.1) port 3000 (#0)
        > GET /openwx/send_friend_message?id=xxxxx&content=hello HTTP/1.1
        > User-Agent: curl/7.29.0
        > Host: 127.0.0.1:3000
        > Accept: */*
        > 
        < HTTP/1.1 200 OK
        < Content-Type: application/json;charset=UTF-8
        < Date: Sun, 13 Dec 2015 04:54:38 GMT
        < Content-Length: 52
        < Server: Mojolicious (Perl)
        <
        * Connection #0 to host 127.0.0.1 left intact
        
        {"status":"发送成功","id":23910327,"code":0}
    
    更多接口参数说明参考[Openwx插件API文档](API.md)
    
3.  *我是一个极客，我只想能够在命令行上通过  IRC 的方式来玩转微信聊天*

    请阅读[IRCShell插件使用步骤](IRC.md)
    
### 核心依赖模块

* [Mojolicious](https://metacpan.org/pod/Mojolicious)
* [Encode::Locale](https://metacpan.org/pod/Encode::Locale)

### 相关文档

* [更新日志](Changes)
* [开发&配置文档](doc/Weixin.pod)
* [基于HTTP的API文档](API.md)
* [FAQ](FAQ.md)

### 官方交流

* [QQ群](http://jq.qq.com/?_wv=1027&k=kjVJzo)
* [IRC](http://irc.perfi.wang/?channel=#Mojo-Webqq)

### 友情链接

*JavaScript*

* [wechaty](https://github.com/zixia/wechaty) Wechaty is wechat for bot in Javascript(ES6). It's a Personal Account Robot Framework/Library.
* [wechatircd](https://github.com/MaskRay/wechatircd) 用IRC客户端控制微信网页版
* [Weixinbot](https://github.com/feit/Weixinbot) Nodejs 封装网页版微信的接口，可编程控制微信消息

*Python*
* [WeixinBot](https://github.com/Urinx/WeixinBot) 网页版微信API，包含终端版微信及微信机器人
* [ItChat](https://github.com/littlecodersh/ItChat) 微信个人号接口、微信机器人及命令行微信。三十行即可自定义个人号机器人

*Java*
* [wechat-client](https://github.com/dadiyang/wechat-client) Mojo-Weixin HTTP 接口 的 Java 客户端，非常感谢 @dadiyang 贡献

### 赞助奖励

可以通过扫描支付宝付款二维码进行赞助，感谢您的支持和鼓励

![donate](screenshot/donate.jpg)

特别致谢 [那些曾经赞助的小伙伴们](https://github.com/sjdy521/Mojo-Thanks)

### COPYRIGHT 和 LICENCE

Copyright (C) 2014 by sjdy521

This library is free software; you can redistribute it and/or modify it under [MIT License](LICENSE) and [Anti 996 License](LICENSE.996ICU)


Changes
2019-09-29 Mojo::Weixin v1.4.6
1)修复Openwx插件错误Can't locate object method "context" via package "Mojo::Weixin::Log"

2019-04-10 Mojo::Weixin v1.4.5
1）修复微信聊天过程中遇到"对方正在输入"提示时，接收消息接口响应异常（"Bad response start-line"），引起程序终止的严重bug
2）增加http请求发生前的debug日志，使得请求建连失败的也可以知道是什么请求
3）增加Openwx插件poll_api心跳请求debug日志
4）post_api接口返回json格式无效时给出原因提示
5）增加Anti 996ICU License

2019-04-08 Mojo::Weixin v1.4.4
1)Openwx插件增加post_message_filter参数来过滤上报的消息内容
2）文档细节完善
3）祝自己生日快乐:-)

2018-11-22 Mojo::Weixin v1.4.3
1）修复Mojo::Transaction::success is DEPRECATED的问题
2）Mojolicious最低依赖版本提升至v8.02

2018-10-09 Mojo::Weixin v1.4.2
1）IO::Socket::SSL依赖模块版本最低2.009
2）Mojolicious依赖模块版本最低7.83

2018-06-29 Mojo::Weixin v1.4.1
1）保存的媒体文件权限设置为0644，修复保持文件时，修改权限调用chmod windows跨平台不兼容问题
2）小冰公众号原本的account： xiaoice-ms 已经失效了，现在只能使用名称来找到 
   修复了因此造成的 XiaoiceReplay 插件 以及 /openwx/consult 接口存在的问题
3）修复/openwx/consult接口返回的消息错乱问题
4）Openwx插件跨域支持完善
5）dockfile更新 解决docker镜像无法制作问题
6）更新依赖关系，解决 IO::Socket::SSL 2.009+ required for TLS support 问题

2017-04-17 Mojo::Weixin v1.4.0
1）Dockerfile更新，换用centos基础镜像，体积更小
2）修复Beauty插件的bug
3）下载的文件使用原始文件名称保存
4）插件IRCShell使用sendFile指令增加media_dir参数

2017-02-26 Mojo::Weixin v1.3.9
1）开放HTTP的Agent自定义属性
2）修复重复发消息的BUG，每个消息对象增加一个uid属性，作为消息唯一标识，再多次失败尝试过程中，服务端对相同uid不会重复
3）修复MsgType为51时的bug 感谢 @cntlis
4）修复账号主域名非wx.qq.com时无法正确识别的bug
5）无法初始化好友和群信息时，直接返回空数据，避免程序退出
6）新增转账消息支持
7）默认日志打印到STDERR
8）win系统默认禁用终端颜色显示
9）增加公告系统
10）Openwx插件一系列改进：
    增加 /openwx/get_group_basic_info 接口 参考issue： https://github.com/sjdy521/Mojo-Weixin/issues/123
    新增post_stdout参数，上报数据是否打印到stdout，适合管道交互信息方式，默认0
    插件端口监听失败会终止程序运行，添加监听端口日志，方便定位问题
    事件上报添加登录失败事件
    修复上报接口直接回复消息时，返回json格式中文包含\uxxxx转义形式时会报错的问题

2017-11-08 Mojo::Weixin v1.3.8
1）新增主动撤回消息功能 感谢 @呸，管杀不管埋
2）文档错误纠正 感谢 @新杰

2017-08-31 Mojo::Weixin v1.3.7
1）修复controller运行过程中报max_client异常的bug

2017-08-30 Mojo::Weixin v1.3.6
1）Openwx上报好友消息增加sender_category属性，用于区分发送者是好友还是公众号
2）初始化好友信息时，如果好友中没有"文件传输助手"，则主动增加这样一个好友，方便做发送消息测试
3）接收消息（synccheck）请求异常时，会延迟10s再次检查消息，这个时间太长影响消息正常接收，改为3s 感谢 @Yeso 反馈
4）增加controller状态检查，当controller退出后，所有client也自动退出
5）修复linux环境下无法通过/openwx/stop_client接口来终止程序的bug（升级后请手动清除模板文件） 感谢 @Aspirin 协助
6）修复 /openwx/stop_client 返回结果不正确的问题
7）controller增加max_clients 参数，默认允许创建的最大客户端数量100

2017-07-03 Mojo::Weixin v1.3.5
1）对于接收消息中无法识别的消息代码，打印日志中增加具体的代码值，同时，客户端会触发 unknown_retcode事件
2）修复由于正则匹配不严谨导致在某些情况下，无法正确获取群成员id的bug
3）new中增加send_interval参数可以自定义全局发送消息时间间隔
4）Openwx插件中 /openwx/send_friend_message 和 /openwx/send_group_message接口增加async=1来实现异步发送消息

2017-05-11 Mojo::Weixin v1.3.4
1）修复Openwx中存在的编码转换问题
2）获取通讯录信息接口调用 POST 变更为 GET
3）new中新增 is_update_all_friend 参数，控制是否获取全部的好友数据
  当前发现分页获取全部好友数据可能会存在问题
  增加这样一个调试参数，当获取通讯录信息异常时，可以收到将该参数设置为0,只获取第一页的好友数据
4）支持接收和自动下载接收到的文件

2017-04-14 Mojo::Weixin v1.3.3
1）修复微信confirming状态bug (https://github.com/sjdy521/Mojo-Weixin/issues/73)
2）Openwx插件 增加 跨域响应头 Access-Control-Allow-Origin: *
3）XiaoiceReply插件产生的智能回复消息 $msg->from 属性设置为"bot"
4）state.json文件中增加previous_state属性，记录当前状态的前一个状态
5）新增MiPush/HwPush/FmPush插件，支持小米手机、华为手机、魅族手机消息推送 感谢 @heipidage
6）修复ShowQRCodeInTerm不可用问题

2017-02-14 Mojo::Weixin v1.3.2
1）GCM插件增加is_ban_official参数用于控制是否推送公众号消息
2）IRCShel插件完善，非常感谢 @eleven @csrgxtu 等人在整个过程中的支持
    使用sendFile XXX 或者 !!XXX 两种形式发送文件
    load_friend参数默认为0
    考虑到个人隐私安全upload_api默认不开启
    文件上传图床，由于https://sm.ms/api/upload接口不支持图片以外的文件，换成兼容http://img.vim-cn.com/的接口形式
    新增auto_join_channel参数，默认为1，收到未加入的频道消息会自动加入频道并接收消息，主要为了防止漏收
    修复HexChat客户端使用/list无法获取到人数为空频道信息
    修复用户更新昵称中存在空格等特殊字符，导致irc虚拟用户异常的问题
    增加allow_group/ban_group参数
    增加登录验证支持
    增加ssl加密支持
3）Openqq插件增加https支持

2017-02-08 Mojo::Weixin v1.3.1
1）IRCShell插件完善
   /query xxxx 如果用户不存在时，会实时从好友中搜索并创建用户
   设置了upload_api时，url地址会直接替换原始的本地文件路径
2）修复产生 "Mojo::Util::slurp is DEPRECATED" 的警告
3）修复分片上传分片处理上的错误导致上传文件不正确或无法上传文件问题
4）改进了判断消息是否为接收消息的判断逻辑，兼容更多服务端数据异常的情况
5）synccheck/sync时采用的实际上是两个独立的key：syncCheckKey和syncKey
   程序原来是混为一起，可能导致重复接收消息或接收消息异常，现已修复
6）GCM插件添加发送者类别支持回复功能
7）修复在开启事件上报之后短时内内重启程序，无法重用之前的登录cookie实现免扫码登录的bug
8）修复controller设置了log_path的情况下，创建的客户端运行日志无法打印到log_path指定路径的bug
9）Openwx插件相关改进
   新增好友验证事件 friend_request 上报
   新增/openwx/accept_friend_request 接口用于接受好友请求

2017-02-03 Mojo::Weixin v1.3.0
1）修复/openwx/get_user_info接口返回的个人信息中account属性为空的bug 感谢 @超 反馈
2）移植 PostQRCodeToTelegram插件到Weixin
3）新增 is_load_plugin 方法 检测客户端是否加载某个插件
4）GCM插件支持推送事件，修复GCM插件二维码url错误,无视GCM插件设置allow_group=>[]的情况
5) IRCShell插件完善
    插件参数master_irc_user更名为master_irc_nick
    插件参数image_api更名为upload_api
    上传api替换为https://sm.ms/api/upload接口
    新增群组、群组名称修改irc上也能够自动同步
    修复irc无法正常发送好友、群组消息的问题
6）完善MakeFile.PL文件

2017-01-16 Mojo::Weixin v1.2.9
1）新增GCM插件
2）修复/openwx/create_group创建的群组，上报new_group事件时群组id为空的bug 感谢 @霸牙 反馈
3）Dockerfile完善，充分利用cache机制，加快build速度，设置镜像为北京时区

2017-01-14 Mojo::Weixin v1.2.8
1）对于请求失败重试，debug日志增加了n/m的重试过程，例如:
[17/01/11 10:08:25] [warn] http://127.0.0.1:80/ 请求(0/2)失败: 404 Not Found
[17/01/11 10:08:25] [warn] http://127.0.0.1:80/ 请求(1/2)失败: 404 Not Found
[17/01/11 10:08:25] [warn] http://127.0.0.1:80/ 请求(2/2)失败: 404 Not Found
2）减少Openwx插件事件上报失败的重试次数
3）设置$client->log_unicode 为0 避免某些情况下产生的乱码问题
4）修复Openwx插件如下api接口不可用问题，感谢 @霸牙 反馈
/openwx/create_group
/openwx/invite_friend
/openwx/kick_group_member
5）基于Docker Hub实现自动化集成

2017-01-10 Mojo::Weixin v1.2.7
1）修复XiaoiceReply插件无法转发文件图片的bug
   XiaoiceReply插件增加allow_friend_reply/allow_group_reply参数
2）修复二维码邮件乱码
3）移除非开源推荐
4）加载了UploadQRcode插件之后，Openwx插件input_qrcode事件上报也会带上上传的二维码url
5）压缩或删除一些图片，减少这个项目的体积
6）Openwx插件auth认证失败时返回json格式数据而非原来的文本错误信息
7）修复客户端正常调用->stop() 退出时无法清理临时文件的bug
   修复客户端正常调用->stop() 退出时无法正常上报事件的bug
   修复ctrl+c时客户端报错 Mojo::IOLoop already running at /usr/local/share/perl5/Mojo/UserAgent.pm line 60

2016-12-29 Mojo::Weixin v1.2.6
1）修复由于官方改对导致误发正常发送视频文件的bug
   文件最大不能超过20M，大于512KB会采用分片上传机制，每个分片大小512K
2）增加 before_send_message 事件，方便全局设置发送消息的小尾巴
3）new中新增属性 is_init_group_member，控制是否在客户端初始登录时获取详细的群成员信息（比如account/sex）等
   默认为0，如果你很在意这些群成员属性，可以设置为1，但可能会极大降低启动速度
4）修复 AutoVerify 插件不可用的问题，感谢 @Hector.X 的反馈
5）UploadQRcode插件采用新的https://sm.ms/的图床接口，感谢 @那谁 的建议
6）修复在多账号Controller控制下，/openwx/check_event接口无法正常工作的问题

Makefile.PL
use 5.010001;
use ExtUtils::MakeMaker "6.46";
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'Mojo::Weixin',
    VERSION_FROM      => 'lib/Mojo/Weixin.pm', # finds $VERSION
    DISTNAME            => 'Mojo-Weixin',
    LICENSE             => "perl",
    PREREQ_PM         => {
        "Compress::Raw::Zlib"   => 0,
        "IO::Compress::Gzip"    => 0,
        "Time::HiRes"           => 0,
        "Time::Piece"           => 0,
        "Time::Seconds"         => 0,
        "Digest::SHA"           => 0,
        "Digest::MD5"           => 0,
        "Encode::Locale"        => 0,
        "IO::Socket::SSL"       => '2.009',
        "Mojolicious"           => '8.02',
    }, # e.g., Module::Name => 1.1
    dist              => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => 'Mojo-Weixin-* MANIFEST' },
    META_MERGE => {
        'meta-spec' => { version => 2 },
        resources => {     
            repository=>{
                type    => 'git',
                url     => 'git://github.com/sjdy521/Mojo-Weixin.git',
                web     =>  'https://github.com/sjdy521/Mojo-Weixin',
            },
        },
    },
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (
       #ABSTRACT_FROM  => 'lib/Webqq/Client.pm', # retrieve abstract from module
       ABSTRACT       => 'A Weixin Client Framework base on Mojolicious',
       AUTHOR         => 'sjdy521 <sjdy521@163.com>') : ()),
);
